File: LinkedListCycleDetection.java
class Solution {
    public static void main(String[] args) {
        ListNode cycListNode = Utility.createCycleListNode();
        ListNode norListNode = Utility.createSampleListNode();

        System.out.println(hasCycle(norListNode));
        System.out.println(hasCycle(cycListNode));
    }

    public static boolean hasCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;

            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
}

File: linked_list_cycle_detection.py
from typing import Optional
from list_node import ListNode
from list_node_utils import ListNodeUtils


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
            if (slow == fast):
                return True
        
        return False
    
cycle_linked_list = ListNodeUtils.create_cycle_list([1, 2, 3, 4], 1)
solution = Solution()
print(solution.hasCycle(cycle_linked_list))
        